app/core/tsa_batch_job.py

```
@@ -0,0 +1,122 @@
+                try:
+                    # Deterministic batch hash: sort by event_id then hash and concatenate.
+                    sorted_items = sorted(
+                        tenant_items, key=lambda it: (it.event_id, it.payload_hash_hex)
+                    )
+                    payload = "".join(it.payload_hash_hex for it in sorted_items).encode(
+                        "ascii"
+                    )
+                    batch_hash = hashlib.sha256(payload).hexdigest()
+
+                    async with AsyncSessionLocal() as db:
+                        async with db.begin():
```

Ruff (0.15.5)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)